File: backend/arkham-api/library/Arkham/Taboo.hs
module Arkham.Taboo (module Arkham.Taboo, module Arkham.Taboo.Types) where

import Arkham.Card
import Arkham.CommitRestriction
import Arkham.Customization
import Arkham.Matcher
import Arkham.Prelude
import Arkham.Taboo.Types
import Data.Map.Strict qualified as Map
import GHC.Records

tabooListModify :: TabooList -> CardDef -> CardDef
tabooListModify = \case
  TabooList15 -> tabooList15Modify
  TabooList16 -> tabooList16Modify
  TabooList18 -> tabooList18Modify
  TabooList19 -> tabooList19Modify
  TabooList20 -> tabooList20Modify
  TabooList21 -> tabooList21Modify
  TabooList22 -> tabooList22Modify
  TabooList23 -> tabooList23Modify

tabooList15Modify :: CardDef -> CardDef
tabooList15Modify cdef = case toCardCode cdef of
  "02266" -> cdef {cdLimits = [MaxPerRound 1]} -- Ace in the Hole
  "02229" -> cdef {cdCommitRestrictions = [MaxOnePerTest]} -- Quick Thinking
  "03315" -> cdef {cdExceptional = True} -- Key of Ys
  _ -> cdef

tabooList16Modify :: CardDef -> CardDef
tabooList16Modify cdef = case toCardCode cdef of
  "02229" -> cdef {cdCommitRestrictions = [MaxOnePerTest]} -- Quick Thinking
  _ -> tabooList15Modify cdef

tabooList18Modify :: CardDef -> CardDef
tabooList18Modify cdef = case toCardCode cdef of
  "02229" -> cdef -- Quick Thinking, removed restriction
  _ -> tabooList16Modify cdef

tabooList19Modify :: CardDef -> CardDef
tabooList19Modify cdef = case toCardCode cdef of
  _ -> tabooList18Modify cdef

tabooList20Modify :: CardDef -> CardDef
tabooList20Modify cdef = case toCardCode cdef of
  "04110" ->
    cdef
      { cdFastWindow =
          Just
            $ RevealChaosToken #when You
            $ IsSymbol
            <> ChaosTokenFaceIsNot #autofail
      }
  "07268" -> cdef {cdExceptional = False, cdDeckRestrictions = [PerDeckLimit 1]}
  "08055" -> cdef {cdLimits = [MaxPerRound 1]}
  _ -> tabooList19Modify cdef

tabooList21Modify :: CardDef -> CardDef
tabooList21Modify cdef = case toCardCode cdef of
  "08076" -> cdef {cdLimits = [MaxPerGame 2]}
  "09022" -> cdef {cdCustomizations = Map.adjust (+ 1) InscriptionOfTheHunt cdef.customizations}
  _ -> tabooList20Modify cdef

tabooList22Modify :: CardDef -> CardDef
tabooList22Modify cdef = case toCardCode cdef of
  _ -> tabooList21Modify cdef

tabooList23Modify :: CardDef -> CardDef
tabooList23Modify cdef = case toCardCode cdef of
  "08113" -> cdef {cdDeckRestrictions = [CampaignModeOnly]}
  _ -> tabooList22Modify cdef

tabooed :: HasField "taboo" a (Maybe TabooList) => TabooList -> a -> Bool
tabooed tbl = maybe False (>= tbl) . getField @"taboo"

File: frontend/src/arkham/taboo.ts
import { TabooList } from '@/arkham/types/TabooList'

export function displayTabooList(taboo: TabooList): string {
  switch (taboo) {
    case "TabooList15": return "1.5 (Apr 23, 2019)"
    case "TabooList16": return "1.6 (Sep 27, 2019)"
    case "TabooList18": return "1.8 (Oct 15, 2020)"
    case "TabooList19": return "1.9 (Jun 28, 2021)"
    case "TabooList20": return "2.0 (Aug 26, 2022)"
    case "TabooList21": return "2.1 (Aug 30, 2023)"
    case "TabooList22": return "2.2 (Feb 20, 2024)"
    case "TabooList23": return "2.3 (Oct 24, 2024)"
    case "TabooList24": return "2.4 (Jul 11, 2025)"
    default: return "Unknown Taboo List"
  }
}

function tabooListIdToTabooList(tabooId: number): TabooList | null {
  switch (tabooId) {
    case 1: return "TabooList15"
    case 2: return "TabooList16"
    case 3: return "TabooList18"
    case 4: return "TabooList19"
    case 5: return "TabooList20"
    case 6: return "TabooList21"
    case 7: return "TabooList22"
    case 8: return "TabooList23"
    case 9: return "TabooList24"
    default: return null
  }
}

export function displayTabooId(tabooId: number): string {
  const taboo = tabooListIdToTabooList(tabooId)
  return taboo ? displayTabooList(taboo) : "Unknown Taboo List"
}
